Feature: generate man pages

mohamedawnallah opened the pull request:
## Change Description
Generate man pages automatically for both `lncli` and `lnd` commands and move the artifacts man pages `lncli.1` and `lnd.1` to `/usr/local/share/man/man1` directory that process happens while running `make install` command.

Fixes #5605

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

ellemouton commented on cmd/lncli/commands.go:
```diff
@@ -1400,6 +1402,38 @@ func channelBalance(ctx *cli.Context) error {
+
+	// Generate the man pages for lnd as lnd.1.
+	config := lnd.DefaultConfig()
+	fileParser := flags.NewParser(&config, flags.Default)
```
I think we need to overwrite the parser's name here (`fileParser.Name = "lnd"`) Else `man lnd` spits out:

```
lncli(1)                  General Commands Manual                  lncli(1)

NAME
       lncli -

```

mohamedawnallah replied:
Thanks! Addressed

ellemouton commented on cmd/lncli/commands.go:
```diff
@@ -1400,6 +1402,38 @@ func channelBalance(ctx *cli.Context) error {
+	fileParser := flags.NewParser(&config, flags.Default)
+	var buf bytes.Buffer
+	fileParser.WriteManPage(&buf)
+	err = os.WriteFile("lnd.1", buf.Bytes(), 0644)
```
formatting nit: can we space things out here a bit pls?

```
        config := lnd.DefaultConfig()
	fileParser := flags.NewParser(&config, flags.Default)
	fileParser.Name = "lnd"
	
	var buf bytes.Buffer
	fileParser.WriteManPage(&buf)
	
	err = os.WriteFile("lnd.1", buf.Bytes(), 0644)
	if err != nil {
		return err
	}

	return nil
```

mohamedawnallah replied:
Thanks! Addressed

ellemouton commented on Makefile:
```diff
@@ -122,11 +124,21 @@ build-itest-race:
 	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lnd
 	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lncli
+	# Invoke the 'manpages' target after installing binaries.
+	make manpages
```
ok so the one thing here is that since our cli commands contain lots of examples that include "< ... >", the `go-md2man` tool sees these as HTML span tags and so this command then spits out a bunch of these warnings:

`WARNING: go-md2man does not handle node type HTMLSpan`

(i've confirmed that this is the issue cause I've tested removing all comments with `< .. >` & this does resolve the issue)

So I think we should figure out a way to get rid of this warning. Either we need to convert all the `<..>` to `[...]` but this might be confusing to a user - they might think that the square brackets are required.... Other option is to find some way to swallow this warning.

mohamedawnallah replied:
I’ve done some research on this here are my two cents:

- Currently, the code related to detecting HTML blocks in the `go-md2man` tool is located in [`roff.go`](https://github.com/cpuguy83/go-md2man/blob/c2c0656178869720a1d9b013ed881db8b7a191ae/md2man/roff.go#L181-L185) file and is used by `github.com/urfave/cli` package when running `ToMan()` and swallowing these warnings requires a change to `go-md2man` which may not feasible in the meantime.
- I've found this warning on a few GitHub issues e.g this issue on [spf13/cobra](https://github.com/spf13/cobra/issues/1137) package and the way to be solved is to replace `<...>` with `[...]` as you said also he put this IBM reference [Reading command-line syntax](https://www.ibm.com/docs/en/iis/8.5?topic=home-reading-command-line-syntax) and also many issues solved this problem by this replacement.

ellemouton replied:
im not sure we should switch to `[]` because usually when we have `[]` we want the user to actually put the `[]` characters in the command. Whereas with `<>` it is implied that these should not be in the final command. 

I still think you can swallow the warning with some bash trickery. 
Something something:

```
your-command 2>&1 | grep -v "go-md2man does not handle node type HTMLSpan" || true
```

mohamedawnallah replied:
Aha thanks a lot, @ellemouton I thought swallowing it on the package level not bash level 🙏

mohamedawnallah replied:
Thanks. Addressed!

ellemouton commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -188,6 +188,9 @@
    to `closechannel` cmd](https://github.com/lightningnetwork/lnd/pull/8350).
 
+* Generate man pages automatically using `lncli generatemanpage` command for
+  both `lncli` and `lnd` commands when running `make install` in the Makefile (https://github.com/lightningnetwork/lnd/pull/8525).
```
nit: see how the linking is done in other entries here. Make some part of the text clickable instead of just giving the raw link

mohamedawnallah replied:
Thanks! Addressed

ellemouton reviewed (commented):
Nice! thanks for this. 

Btw - maybe add to the PR description that this replaces https://github.com/lightningnetwork/lnd/pull/7424

ellemouton reviewed (commented):
see my comment re swallowing the error warning with some bash trickery

mohamedawnallah commented:
Thanks again @ellemouton for your review. I've addressed all your comments!

ellemouton reviewed (commented):
Thanks @mohamedawnallah ,

I think we should instead put the script in a dedicated bash script & then call it from the makefile. That way we can reduce the logs printed out too. 

So I suggest converting the makefile method to:

```
manpages:
	@$(call print, "Generating man pages lncli.1 and lnd.1.")
	./scripts/gen_man_pages.sh $(DESTDIR) $(PREFIX)
```

and then add a `gen_man_pages.sh` file to the `scripts` folder with the following contents:

```
#!/bin/bash

# Usage: ./gen_man_pages.sh DESTDIR PREFIX

DESTDIR="$1"
PREFIX="$2"

# Check if lncli is installed.
if ! command -v lncli &> /dev/null
then
    echo "lncli could not be found. Please install lncli before running this script."
    exit 1
fi

# Ignore warnings regarding HTMLBlock detection in go-md2man package
# since using "<...>" is part of our docs.
lncli generatemanpage 2>&1 | grep -v "go-md2man does not handle node type HTMLSpan" || true

echo "Installing man pages to $DESTDIR$PREFIX/share/man/man1."
install -m 644 lnd.1 "$DESTDIR$PREFIX/share/man/man1/lnd.1"
install -m 644 lncli.1 "$DESTDIR$PREFIX/share/man/man1/lncli.1"
```

let me know what you think :)

ellemouton commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -194,6 +194,9 @@
 * The [`closeallchannels` command now asks for confirmation before closing 
   all channels](https://github.com/lightningnetwork/lnd/pull/8526).
 
+* [Generate man pages automatically using `lncli generatemanpage` command
+  for both `lncli` and `lnd` commands](https://github.com/lightningnetwork/lnd/pull/8525) when running `make install` in the Makefile.
```
pls try wrap this to 80 chars 🙏

mohamedawnallah replied:
Thanks! Addressed

mohamedawnallah commented:
Thanks @ellemouton for your feedback. I agree with isolating the man pages script into an isolated bash script it is a much better solution. I've addressed all of your feedback!

ellemouton reviewed (changes requested):
@mohamedawnallah  - did you test this? 

```
make manpages
 Generating man pages lncli.1 and lnd.1.
./scripts/gen_man_pages.sh  /usr/local
make: ./scripts/gen_man_pages.sh: Permission denied
make: *** [manpages] Error 1
```

you need to make sure that the new bash script file is executable.

In general, pls always make sure to tests your changes & also to wait for the CI to be green (besides some flakes) before re-requesting review 🙏 😊

mohamedawnallah commented:
> In general, pls always make sure to tests your changes & also to wait for the CI to be green (besides some flakes) before re-requesting review 🙏 😊

Acknowledged 👍

ellemouton commented on Makefile:
```diff
@@ -122,11 +124,17 @@ build-itest-race:
 	@$(call print, "Installing lnd and lncli.")
 	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lnd
 	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lncli
+	# Invoke the 'manpages' target after installing binaries.
```
i think let's remove this comment since it gets printed out. The bash script already prints out enough info i think :)

mohamedawnallah replied:
SGTM. Updated 👍

ellemouton reviewed (approved):
LGTM! Great work @mohamedawnallah 🙏 

One last micro nit :) otherwise - LGTM

guggero commented on Makefile:
```diff
@@ -122,11 +124,16 @@ build-itest-race:
 	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lncli
+	make manpages
+
+manpages:
```
I think calling `make` from within `make` is a bit weird. How about we do it this way instead to have proper dependencies between build goals:
```
#? install-binaries: Build and install lnd and lncli binaries, place them in $GOPATH/bin.
install-binaries:
	@$(call print, "Installing lnd and lncli.")
	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lnd
	$(GOINSTALL) -tags="${tags}" -ldflags="$(RELEASE_LDFLAGS)" $(PKG)/cmd/lncli

#? manpages: generate and install man pages.
manpages:
	@$(call print, "Generating man pages lncli.1 and lnd.1.")
	./scripts/gen_man_pages.sh $(DESTDIR) $(PREFIX)

#? install: Build and install lnd and lncli binaries, place them in $GOPATH/bin, generate and install man pages.
install: install-binaries manpages
```

mohamedawnallah replied:
Very nice! thanks for this also thanks for adding the help comment above the make target forgot to added it 👍

guggero reviewed (commented):
Very nice. Looks good to me, one comment about the `Makefile`.

guggero commented on Makefile:
```diff
@@ -122,12 +124,20 @@ build-itest-race:
-#? install: Build and install lnd and lncli binaries, place them in $GOPATH/bin
-install:
+#? install: Build and install lnd and lncli binaries, place them in $GOPATH/bin.
+install-binaries:
```
nit: the help hint for this make goal is now outdated. Also, looking at all other help hints, we don't use full stops at the end of sentences everywhere else, so let's make this and the ones below consistent with the rest of the file.

guggero reviewed (approved):
Final nit, otherwise LGTM :tada:
